src/observers/qos_correction.cpp

```diff
@@ -2,72 +2,126 @@
+#include "serenity/resource_helper.hpp"
 
 namespace mesos {
 namespace serenity {
 
 QoSCorrectionObserver::~QoSCorrectionObserver() {}
 
+Try<Nothing> QoSCorrectionObserver::qosCorrectionStateMachine() {
+  if (contentions.get().empty() ||
+      ResourceUsageHelper::getRevocableExecutors(usage.get()).empty()) {
+    noContentionsReceived();
+    // Produce empty corrections and contentions
+    produceResultsAndClearConsumedData();
+    return Nothing();
+  }
 
-Try<Nothing> QoSCorrectionObserver::_syncConsume(
-  const std::vector<Contentions> products) {
-  Contentions newContentions;
-
-  for (Contentions contentions : products)
-    for (Contention contention : contentions)
-      newContentions.push_back(contention);
+  if (iterationCooldownCounter.isSome()) {
+    cooldownPhase();
+    // Produce empty corrections and contentions
+    produceResultsAndClearConsumedData();
+    return Nothing();
+  }
 
-  this->currentContentions = newContentions;
-  this->currentContentions.get().sort(compareSeverity);
+  Try<QoSCorrections> corrections = newContentionReceived();
+  if (corrections.isError()) {
+    // Produce empty corrections and contentions
+    produceResultsAndClearConsumedData();
+    return Error(corrections.error());
+  }
 
-  if (this->currentUsage.isSome()) {
-    this->correctAgent();
+  if (corrections.get().empty()) {
+    // Strategy didn't found aggressors.
+    // Passing contentions to next QoS Controller.
+    produceResultsAndClearConsumedData(QoSCorrections(),
+                                       this->contentions.get());
+    return Nothing();
   }
 
+  // Strategy has pointed aggressors, so don't pass
+  // current contentions to next QoS Controller.
+  produceResultsAndClearConsumedData(corrections.get(),
+                                     Contentions());
   return Nothing();
 }
 
+void QoSCorrectionObserver::produceResultsAndClearConsumedData(
+    QoSCorrections _qosCorrections,
+    Contentions _contentions) {
+  produce<QoSCorrections>(_qosCorrections);
+  produce<Contentions>(_contentions);
 
-Try<Nothing> QoSCorrectionObserver::consume(const ResourceUsage& usage) {
-  this->currentUsage = usage;
+  this->contentions = None();
+  this->usage = None();
+}
 
-  if (this->currentContentions.isSome()) {
-    this->correctAgent();
+void QoSCorrectionObserver::noContentionsReceived() {
+  // Restart state of QoSCorrection observer
+  if (iterationCooldownCounter.isSome()) {
+    iterationCooldownCounter = None();
+  }
+  if (estimatorPipelineDisabled) {
+    StaticEventBus::publishOversubscriptionCtrlEvent(true);
+    estimatorPipelineDisabled = false;
   }
+}
 
-  return Nothing();
+void QoSCorrectionObserver::cooldownPhase() {
+  // Cooldown phase - waiting for system to stabilize.
+  iterationCooldownCounter.get() -= 1;
+  if (iterationCooldownCounter.get() <= 0) {
+    iterationCooldownCounter = None();
+  }
+  return;
 }
 
+Try<QoSCorrections> QoSCorrectionObserver::newContentionReceived() {
```

Comment:
Let's place 'newContentionReceived` and `emptyContentionReceived` near themselves for the convenience. (: